ark/jsonschema/object.ts

```diff
@@ -0,0 +1,336 @@
+				ctx.reject({
+					expected: "an array of unique strings",
+					path: ["required"]
+				})
+			}
+
+			for (const key of jsonSchema.required) {
+				if (key in jsonSchema.properties) requiredKeys.push(key)
+				else {
+					ctx.reject({
+						actual: key,
+						expected: "a key in the 'properties' object",
+						path: ["required"]
+					})
+				}
+			}
+			for (const key in jsonSchema.properties)
+				if (!jsonSchema.required.includes(key)) optionalKeys.push(key)
+		} else {
+			// If 'required' is not present, all keys are optional
+			optionalKeys.push(...Object.keys(jsonSchema.properties))
+		}
+	} else if ("required" in jsonSchema) {
+		ctx.reject({
+			actual:
+				"an object JSON Schema with 'required' array but no 'properties' object",
+			expected: "a valid object JSON Schema"
+		})
+	}
+
+	return {
+		optionalKeys: optionalKeys.map(key => ({
+			key,
+			value: innerParseJsonSchema.assert(jsonSchema.properties![key]).internal
+		})),
+		requiredKeys: requiredKeys.map(key => ({
+			key,
+			value: innerParseJsonSchema.assert(jsonSchema.properties![key]).internal
+		}))
+	}
+}
+
+const parseAdditionalProperties = (jsonSchema: JsonSchema.ObjectSchema) => {
+	if (!("additionalProperties" in jsonSchema)) return
+
+	const properties = Object.keys(jsonSchema.properties ?? {})
+	const patternProperties = Object.keys(jsonSchema.patternProperties ?? {})
+
+	const additionalPropertiesSchema = jsonSchema.additionalProperties
+	if (additionalPropertiesSchema === true) return
+
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.keys(data).forEach(key => {
+			if (
+				properties.includes(key) ||
+				patternProperties.find(pattern => new RegExp(pattern).test(key))
+			)
+				// Not an additional property, so don't validate here
+				return
+
+			if (additionalPropertiesSchema === false) {
+				errors.push(
+					ctx.reject({
+						message: `property ${key} is an additional property, which the provided schema does not allow`
+					})
+				)
+				return
+			}
+
+			const additionalPropertyValidator = innerParseJsonSchema.assert(
+				additionalPropertiesSchema
+			)
+
+			const value = data[key as keyof typeof data]
+			if (!additionalPropertyValidator.allows(value)) {
+				errors.push(
+					ctx.reject({
+						problem: `property ${key} is an additional property so must adhere to additional property schema of ${additionalPropertyValidator.description} (was ${printable(value)})`
+					})
+				)
+			}
+		})
+		return errors.length === 0
+	}
+}
+
+export const validateJsonSchemaObject = JsonSchema.ObjectSchema.pipe(
+	(jsonSchema, ctx): Type<object> => {
+		const arktypeObjectSchema: Intersection.Schema<object> = {
+			domain: "object"
+		}
+
+		const { requiredKeys, optionalKeys } = parseRequiredAndOptionalKeys(
+			jsonSchema,
+			ctx
+		)
+		arktypeObjectSchema.required = requiredKeys
+		arktypeObjectSchema.optional = optionalKeys
+
+		const predicates: Predicate.Schema[] = [
```

Comment:
There are some utils in `@ark/util` that might help you here like `append` or `conflatenate` that will add an element or array to another array if it's not undefined.

Reply:
Looks like `conflatenateAll` does what I need: 2fe1b7ac